File: src/camera2ros.py
#!/usr/bin/python3

import rospy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

class Camera:
    def __init__(self, index):

        cameraName = f"camera_{index}"
        topicName = f"sky_vision/{cameraName}/image_raw" 
        self.capture = cv2.VideoCapture(index) #openCV camera object. Default camera index is 0
        self.bridge = CvBridge() #bridge type that converts between openCV and ROS images
        self.publisher = rospy.Publisher(topicName, Image, queue_size=1)

        rospy.loginfo(f"Trying to open {cameraName}.")
        while(not self.capture.isOpened()):
            rospy.logwarn(f"{cameraName}: Wasn't able to open the camera. Trying again...")
            rospy.sleep(1)

        rospy.loginfo(f"{cameraName} open!")
        cameraInfo = f"{cameraName} INFO:\nWIDTH: {self.capture.get(cv2.CAP_PROP_FRAME_WIDTH)}\nHEIGHT: {self.capture.get(cv2.CAP_PROP_FRAME_HEIGHT)}\nFPS: {self.capture.get(cv2.CAP_PROP_FPS)}\nFORMAT: {self.capture.get(cv2.CAP_PROP_FORMAT)}\nPUBLISHING AS: bgr8"
        rospy.loginfo(cameraInfo)

    def processFrame(self):

        try:
            check, cvImage = self.capture.read() #check returns if the function was successful or not.
            self.publisher.publish(self.bridge.cv2_to_imgmsg(cvImage, "bgr8"))
        
        except CvBridgeError as e:
            if(not check):
                rospy.logerr("Couldn't read the frame.")
            rospy.logerr(e)

    def __del__(self):
        self.capture.release()

def main():
    rospy.init_node(f"camera")
    index = rospy.get_param(f"~camera/index", 0) #default openCV camera index, could also use -1
    cam = Camera(index)

    while not rospy.is_shutdown():
        cam.processFrame()

if __name__ == '__main__':
    main()

File: src/camera2ros.cpp
#include <opencv2/opencv.hpp>
#include "ros/ros.h"
#include "sensor_msgs/Image.h"
#include "std_msgs/Header.h"
#include "string.h"
#include "ros/console.h"
#include <cv_bridge/cv_bridge.h>
#include <iostream>

using namespace cv;

int main(int argc, char **argv){

  ros::init(argc, argv, "camera");
  ros::NodeHandle nodeHandle("~"); //the "~" makes the handler launch the node as private

  int index = 0; //opencv default camera index
  
  //this param has to be private, as we may launch multiple nodes like this
  ros::param::get("~camera/index", index);

  String topicName = "sky_vision/camera_" + std::to_string(index) + "/image_raw";

  sensor_msgs::Image frameROS; //ros image format
  cv_bridge::CvImage frameBridge; //bridge class
  Mat frameCV; //opencv image format
  std_msgs::Header header; //ros images need a header
  uint32_t counter = 0; //counter used to populate the ros image header

  VideoCapture cap(index);

  while(!cap.isOpened() && ros::ok){
    ROS_ERROR("CAMERA_%d: Wasn't able to open the camera. Trying again...\n", index);
    ros::Duration(1).sleep();
    ros::spinOnce();
  }

  ROS_INFO("CAMERA_%d: Succesfully opened the camera. \n", index);


  ros::Publisher cameraPub = nodeHandle.advertise<sensor_msgs::Image>(topicName, 5);

  while(ros::ok()){

    header.seq = counter;
    header.stamp = ros::Time::now();
    cap.read(frameCV);
    frameBridge = cv_bridge::CvImage(header, sensor_msgs::image_encodings::RGB8, frameCV);
    cameraPub.publish(frameBridge);

    counter++;
    ros::spinOnce();
  }

  return 0;
}